internal/fs/hash.go

```diff
@@ -0,0 +1,145 @@
+	h := sha256.New()
+
+	// "../../../vendor", "github.com/account/library"
+	prefix = filepath.Clean(prefix) + pathSeparator
+	prefixLength := len(prefix)
+	if prefixLength > 0 {
+		prefixLength += len(pathSeparator) // if not empty string, include len of path separator
+	}
+	joined := filepath.Join(prefix, pathname)
+
+	// Initialize a work queue with the os-agnostic cleaned up pathname. Note
+	// that we use `filepath.Clean` rather than `filepath.Abs`, because we don't
+	// want the hash to be based on the absolute pathnames of the specified
+	// directory and contents.
+	pathnameQueue := []string{joined}
+
+	for len(pathnameQueue) > 0 {
+		// NOTE: unshift a pathname from the queue
+		pathname, pathnameQueue = pathnameQueue[0], pathnameQueue[1:]
+
+		fi, er := os.Lstat(pathname)
+		if er != nil {
+			err = errors.Wrap(er, "cannot Lstat")
+			return
+		}
+
+		mode := fi.Mode()
+
+		// Skip special files
+		if mode&skipModes != 0 {
+			continue
+		}
+
+		// NOTE: Write pathname to hash, because hash ought to be as much a
+		// function of the names of the files and directories as their
+		// contents. Added benefit is that even empty directories and symbolic
+		// links will effect final hash value.
+		//
+		// NOTE: Throughout this function, we ignore return values from writing
+		// to the hash, because hash write always returns nil error.
+		_, _ = h.Write([]byte(pathname)[prefixLength:])
+
+		if mode&os.ModeSymlink != 0 {
+			referent, er := os.Readlink(pathname)
+			if er != nil {
+				err = errors.Wrap(er, "cannot Readlink")
+				return
+			}
+			// Write the referent to the hash and proceed to the next pathname
+			// in the queue.
+			_, _ = h.Write([]byte(referent))
+			continue
+		}
+
+		fh, er := os.Open(pathname)
+		if er != nil {
+			err = errors.Wrap(er, "cannot Open")
+			return
+		}
+
+		if fi.IsDir() {
+			childrenNames, er := fh.Readdirnames(0) // 0: read names of all children
+			if er != nil {
+				err = errors.Wrap(er, "cannot Readdirnames")
+				// NOTE: Even if there was an error reading the names of the
+				// directory entries, we still must close file handle for the
+				// open directory before we return. In this case, we simply skip
+				// sorting and adding entry names to the work queue beforehand.
+				childrenNames = nil
+			}
+
+			// NOTE: Sort children names to ensure deterministic ordering of
+			// contents of each directory, ensuring hash remains same even if
+			// operating system returns same values in a different order on
+			// subsequent invocation.
+			sort.Strings(childrenNames)
+
+			for _, childName := range childrenNames {
+				switch childName {
+				case ".", "..", "vendor", ".bzr", ".git", ".hg", ".svn":
+					// skip
+				default:
+					pathnameQueue = append(pathnameQueue, pathname+pathSeparator+childName)
+				}
+			}
+		} else {
+			_, _ = h.Write([]byte(strconv.FormatInt(fi.Size(), 10))) // format file size as base 10 integer
+			_, er = io.Copy(h, fh)
```

Comment:
> Assuming that is agreed, your point about CRLF vs LF translation might reveal quite the complex problem for us to tackle.

yep! i've been worried about this one for a while 😄 

> (Also, I think we could even get away with in-memory translating every file, including binary ones, with impunity here, but this seems like a bizarre solution.)

after poking at this and thinking about it for a while, i think this would actually be fine - a simple scanner that just reads the bytes into memory and mangles them there, effectively treating `\r\n` and `\n` as equivalent. while i wasn't able to find the actual logic in git that does this work, i did at least verify that `git hash-object` on a file produces the same hash when i replace a `\r\n` with a `\n`, so that _appears_ to be the strategy they employ.

the (a?) tricky bit is that we could no longer rely on the file size to be an accurate byte count; we could only know afterwards, after having performed the mangling. i believe that would mean we'd either have to write the file length as a suffix, or keep the entire file's contents in a temporary buffer and then pass it to the hasher after we've reached the end and gotten an accurate count.